File: R/48-rotate-image/main.go
package main

import "fmt"

func rotate(matrix [][]int) {
	for i := 0; i < len(matrix); i++ {
		s, e := 0, len(matrix[0])-1
		for s < e {
			matrix[i][s], matrix[i][e] = matrix[i][e], matrix[i][s]
			s++
			e--
		}
	}
	for i := 0; i < len(matrix); i++ {
		for j := 0; j < len(matrix[0]); j++ {
			if i+j < len(matrix)-1 {
				matrix[i][j], matrix[len(matrix)-1-j][len(matrix)-1-i] = matrix[len(matrix)-1-j][len(matrix)-1-i], matrix[i][j]
			}
		}
	}
}

func main() {
	fmt.Println("a")
}

File: R/48-rotate-image/main.rs
impl Solution {
  pub fn rotate(matrix: &mut Vec<Vec<i32>>) {
    let n = matrix.len();
    (0..n).for_each(|r| {
      let (mut s, mut e) = (0, n - 1);
      while s < e {
        (matrix[r][s], matrix[r][e]) = (matrix[r][e], matrix[r][s]);
        s += 1;
        e -= 1;
      }
    });
    (0..n).for_each(|r| {
      (0..n).for_each(|c| {
        if r + c >= n - 1 {
          return;
        }
        (matrix[r][c], matrix[n - 1 - c][n - 1 - r]) = (matrix[n - 1 - c][n - 1 - r], matrix[r][c]);
      });
    });
  }
}
